src/pubsub/index.js

```diff
@@ -0,0 +1,663 @@
+    } else {
+      return message
+    }
+  }
+
+  // API METHODS
+
+  /**
+   * Get a list of the peer-ids that are subscribed to one topic.
+   * @param {string} topic
+   * @returns {Array<string>}
+   */
+  getSubscribers (topic) {
+    if (!this.started) {
+      throw errcode(new Error('not started yet'), 'ERR_NOT_STARTED_YET')
+    }
+
+    if (!topic || typeof topic !== 'string') {
+      throw errcode(new Error('a string topic must be provided'), 'ERR_NOT_VALID_TOPIC')
+    }
+
+    const peersInTopic = this.topics.get(topic)
+    if (!peersInTopic) {
+      return []
+    }
+    return Array.from(peersInTopic)
+  }
+
+  /**
+   * Publishes messages to all subscribed peers
+   * @override
+   * @param {Array<string>|string} topics
+   * @param {Buffer} message
+   * @returns {Promise<void>}
+   */
+  async publish (topics, message) {
+    if (!this.started) {
+      throw new Error('Pubsub has not started')
+    }
+
+    topics = utils.ensureArray(topics)
+    this.log('publish', topics, message)
+
+    const from = this.peerId.toB58String()
+    let msgObject = {
+      receivedFrom: from,
+      from: from,
+      data: message,
+      seqno: utils.randomSeqno(),
+      topicIDs: topics
+    }
+
+    // Emit to self if I'm interested and emitSelf enabled
+    this.emitSelf && this._emitMessage(msgObject)
+
+    // ensure that any operations performed on the message will include the signature
+    msgObject = await this._buildMessage(msgObject)
+
+    // send to all the other peers
+    await this._publish(msgObject)
+  }
+
+  /**
+   * Overriding the implementation of publish should handle the appropriate algorithms for the publish/subscriber implementation.
+   * For example, a Floodsub implementation might simply publish each message to each topic for every peer
+   * @abstract
+   * @param {InMessage} message
+   * @returns {Promise<void>}
+   *
+   */
+  _publish (message) {
+    throw errcode(new Error('publish must be implemented by the subclass'), 'ERR_NOT_IMPLEMENTED')
+  }
+
+  /**
+   * Subscribes to a given topic.
+   * @abstract
+   * @param {string} topic
+   * @param {function} [handler]
+   * @returns {void}
+   */
+  subscribe (topic, handler) {
+    if (!this.started) {
+      throw new Error('Pubsub has not started')
+    }
+
+    if (!this.subscriptions.has(topic)) {
+      this.subscriptions.add(topic)
+      this.peers.forEach((_, id) => this._sendSubscriptions(id, [topic], true))
+    }
+
+    // Bind provided handler
+    handler && this.on(topic, handler)
+  }
+
+  /**
+   * Unsubscribe from the given topic.
+   * @override
+   * @param {string} topic
+   * @param {function} [handler]
+   * @returns {void}
+   */
+  unsubscribe (topic, handler) {
+    if (!this.started) {
+      throw new Error('FloodSub is not started')
+    }
+
+    if (!handler) {
+      this.removeAllListeners(topic)
+    } else {
+      this.removeListener(topic, handler)
+    }
+
+    if (this.listenerCount(topic) === 0) {
```

Comment:
What is this for?
I thought we would want to delete the topic from `this.subscriptions` no matter what. As well as send unsubscriptions.

Reply:
Imagine you have a webapp that subscribes a specific topic in a UI component and the same topic in a different UI component, both with a handler. If one unmounts, the topic would be unsubscribed and no more messages would be received.
I think it makes sense when people use handlers. If you do not provide an handler, the subscription will be always removed.

Reply:
I see. IMO we should reconsider this interface, or at least provide a separate subscription-only interface. Topic subscription should have the ability to be separated from topic event handling, but combining the two here seems to force this coupling, even if its not desired.

I think the bad behavior is removing all listeners on unsubscribe. Because of this, consumers _must_ couple event handling and subscription, because there is no way to perform unsubscription without editing attached event handlers.

Reply:
Well, I have been thinking about it. now that we expose the router directly by libp2p.pubsub , we can probably just leverage the event emitter api, document that the topics are even emitter events and people just need to use its api. It is a new breaking change, but I think we are in a good position to so it!
We definitely need to have a good migration guide for this

Reply:
We just cannot accept the subscription change event name as a topic.

Reply:
`js-ipfs` also relies on the `handlers`. @jacobheun what are your thoughts on getting rid of them?